### PR TITLE
Add user endpoint tests

### PR DESCRIPTION
**Pull request by hariniczi**

### CODE REVIEW

File: tests/test_api.py [hunk 1]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                headers = {'Content-Type': "application/json"}
+                headers.update(get_auth_header(service_accounts['admin']))
+                url = furl(f'/v1/users/{test["name"]}')
+                query_params = {
+                    'user_id': test['name'],
+                    'status': test['status']
+                }
+                url.add(query_params=query_params)
+                user = User.provision_user(directory, test['name'])
+                if test['status'] == 'disabled':
```
Comment (Bento007) [inline tests/test_api.py]: Do we ever validate that the users was created with status disabled?

File: tests/test_api.py [hunk 2]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+            with self.subTest(test['json_request_body']):
+                data=json.dumps(test['json_request_body'])
+                headers={'Content-Type': "application/json"}
+                headers.update(get_auth_header(service_accounts['admin']))
+                resp = self.app.post('/v1/policies/evaluate', headers=headers, data=data)
+                self.assertEqual(test['response']['code'], resp.status_code)
+                self.assertEqual(test['response']['result'], json.loads(resp.body)['result'])
+
+    def test_put_new_user(self):
+        tests = [
```
Comment (Bento007) [inline tests/test_api.py]: Please add a test with a user name that is > and == 64 characters.

File: tests/test_api.py [hunk 3]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                }
+            }
+        ]
+        for test in tests:
+            with self.subTest(test['json_request_body']):
+                data=json.dumps(test['json_request_body'])
+                headers={'Content-Type': "application/json"}
+                headers.update(get_auth_header(service_accounts['admin']))
+                resp = self.app.put('/v1/users', headers=headers, data=data)
+                self.assertEqual(test['response']['code'], resp.status_code)
```
Comment (Bento007) [inline tests/test_api.py]: Add verification that the user was created with the expected values.

File: tests/test_api.py [hunk 4]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                if test['status'] == 'disabled':
+                    user.enable()
+                resp = self.app.put(url.url, headers=headers)
+                self.assertEqual(test['response']['code'], resp.status_code)
+                resp.raise_for_status()
 
-    @unittest.skip("incomplete")
-    def test_put_group(self):
-        pass
+    def test_put_users_groups(self):
```
Comment (Bento007) [inline tests/test_api.py]: change to test_put_username_groups

File: tests/test_api.py [hunk 5]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                }
+                url.add(query_params=query_params)
+                user = User.provision_user(directory, test['name'])
+                if test['action']=='remove':
+                    user.add_groups(test['json_request_body']['groups'])
+                resp = self.app.put(url.url, headers=headers, data=data)
+                self.assertEqual(test['response']['code'], resp.status_code)
+                resp.raise_for_status()
+
+    def test_get_users_groups(self):
```
Comment (Bento007) [inline tests/test_api.py]: change to test_get_username_groups

File: tests/test_api.py [hunk 6]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+        headers = {'Content-Type': "application/json"}
+        headers.update(get_auth_header(service_accounts['admin']))
+        name = "test_user_group_api@email.com"
+        User.provision_user(directory, name)
+        resp = self.app.get(f'/v1/users/{name}/groups', headers=headers)
+        self.assertEqual(0, len(json.loads(resp.body)['groups']))
+        resp.raise_for_status()
+
+    def test_put_users_roles(self):
```
Comment (Bento007) [inline tests/test_api.py]: change to test_put_username_roles

File: tests/test_api.py (hunk 5 above)
Comment (Bento007) [reply tests/test_api.py]: add a test where the user is in a group or two and groups are returned from the API

File: tests/test_api.py [hunk 7]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                if test['action']=='remove':
+                    user.add_roles(test['json_request_body']['roles'])
+                resp = self.app.put(url.url, headers=headers, data=data)
+                self.assertEqual(test['response']['code'], resp.status_code)
+                resp.raise_for_status()
 
-    @unittest.skip("incomplete")
-    def test_get_group(self):
-        pass
+    def test_get_users_roles(self):
```
[the remark this answers is not in the document]
Comment (Bento007) [reply tests/test_api.py]: add a test where the user has 0, 1, or 2+ roles and they are returned

File: tests/test_api.py [hunk 8]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+        name = "test_user_api@email.com"
+        User.provision_user(directory,name)
+        resp = self.app.get(f'/v1/users/{name}/',headers=headers)
+        self.assertEqual(name,json.loads(resp.body)['name'])
+        resp.raise_for_status()
+
+    def test_put_user(self):
+        tests = [
+            {
+                'name': "test_put_user0@email.com",
```
[the remark this answers is not in the document]
Comment (hariniczi) [reply tests/test_api.py]: I chose this method since I access name and status in other places too.

File: tests/test_api.py (hunk 2 above)
Comment (hariniczi) [reply tests/test_api.py]: Added in the name tests

File: tests/test_api.py [hunk 9]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+                    "roles": [Role.create(directory, "role_05").name],
+                    "policy": create_test_statement("policy_05")
+                },
+                'response': {
+                    'code': 400,
+                    'result': True
+                }
+            }
+        ]
+        for test in tests:
```
Comment (Bento007) [inline tests/test_api.py]: add a test to to put a user that already exists.

File: tests/test_api.py [hunk 10]
```diff
@@ -167,58 +167,263 @@ class TestApi(unittest.TestCase):
...
+            },
+            {
+                'json_request_body': {
+                    "groups": [Group.create(directory, "group_05").name],
+                    "roles": [Role.create(directory, "role_05").name],
+                    "policy": create_test_statement("policy_05")
+                },
+                'response': {
+                    'code': 400,
+                    'result': True
```
Comment (Bento007) [inline tests/test_api.py]: This doesn't appear to be used anywhere. Please remove or otherwise use it

File: tests/test_api.py (hunk 2 above)
Comment (Bento007) [reply tests/test_api.py]: move these tests to "test_user_api"

File: tests/test_user_api.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+from tests.common import get_auth_header, service_accounts, create_test_statement
+import fusillade
+from fusillade import directory, User
+from fusillade.clouddirectory import cleanup_directory, User, Group, Role
+
+from tests.infra.server import ChaliceTestHarness
+# ChaliceTestHarness must be imported after FUSILLADE_DIR has be set
+
+
+def setUpModule():
```
Comment (Bento007) [inline tests/test_user_api.py]: if setup is not required, remove it

File: tests/test_api.py (hunk 2 above)
Comment (hariniczi) [reply tests/test_api.py]: done

File: tests/test_user_api.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+                query_params = {
+                    'user_id': test['name'],
+                    'action': test['action']
+                }
+                url.add(query_params=query_params)
+                user = User.provision_user(directory, test['name'])
+                if test['action'] == 'remove':
+                    user.add_roles(test['json_request_body']['roles'])
+                resp = self.app.put(url.url, headers=headers, data=data)
+                self.assertEqual(test['response']['code'], resp.status_code)
```
Comment (Bento007) [inline tests/test_user_api.py]: if you're checking the response code here you don't need `resp.raise_for_status()` below

**Comment by jharini** [comment]
Reassigning to myself and closing it

**Pull request merged**